### ISSUE TITLE
INSPQ data are aggressively cached, leading to old data being served

### ISSUE THREAD

**Issue by jeanpaulrsoucy** [ccodwg/Covid19CanadaArchive#185]
I was tracking down the cause of an issue in the [processing](https://github.com/ccodwg/Covid19CanadaDataProcess/issues/7) of INSPQ data. I loaded it up in R and it turned out the [link](https://www.inspq.qc.ca/sites/default/files/covid/donnees/covid19-hist.csv) was serving an old version of the dataset (max date was 2021-06-28). Adding a `randNum` parameter to the end of the URL gave the correct max date (2021-07-12).

@mschoettle, I think you mentioned this potential issue a while back, but it appears I have run into it. Some of the data in the archive is likely tainted due to this. Oh well...lesson learned. I'll do a more thorough inventory of this when I have time.

To start, notes of this should be made in the metadata of any INSPQ datasets where the original URL may have contained a `randNum` parameter.

I will also have to figure out how to deal with this in `archiver.py`. Is appending a random number with an option sufficient? Is it safe to perform this on *all* links by default? Or can this be achieved using [HTTP headers](https://softwareengineering.stackexchange.com/a/371008)?

**Comment by jeanpaulrsoucy** [comment]
I can add a "no-cache" policy to the header, but this is not always respected (I just checked with an INSPQ dataset). However, appending `?randNum=xxxx` is not safe either, since some URLs (particularly with ESRI dashboards), already contain parameters, so the appropriate suffix would be with `&` instead. For now, I will add the "no-cache" header and make it a dataset-specific option to add a randNum (based on the time of request to ensure uniqueness) to the URL.

**Comment by mschoettle** [comment]
I actually dealt with this problem recently again for INSPQ data. I think it was on July 1st or 2nd. I can most likely provide you with some of the missing datasets. We scrape several of them for the COVID-19 Montreal Dashboard. Just let me know which ones and for which date ranges.

You could also try `Pragma: no-cache` ([this is apparently intended for requests](https://stackoverflow.com/a/15050018)) but I don't think it makes a difference. When I looked into this a while ago setting certain HTTP headers did not work because they use Cloudflare which does not care about them (by default). It also sounds like a Cloudflare configuration issue on INSPQ's part since Cloudflare has [options to configure this](https://stackoverflow.com/a/41269339).

What I ended up doing was to append the unix time to the request and depending on the URL use the correct query parameter separator:

```
# add "random" string to URL to prevent server-side caching
unix_time = datetime.now().strftime('%s')
query_param_separator = '&' if '?' in url else '?'
url = url + query_param_separator + unix_time
```

This method also works with ArcGIS URLs. But I agree that it is better to use the header by default and only use this method for specific datasets.

**Comment by jeanpaulrsoucy** [comment]
Yup, that's the method I ended up going with to fix the data download/processing tool (https://github.com/ccodwg/Covid19CanadaData/commit/fc441986fcdf21dbe371dc8cebcf8f76e9dd133d). It worked great.

I've just adapted the same method for the Python archival tool. Again, works great.

I'll have to read more into whether it's safe to include an extra parameter by default to ensure no caching.

@mschoettle I'll send you an email regarding those archived datasets. I may just ask for everything you have and sort things out at a later date...

Thanks again!

**Comment by jeanpaulrsoucy** [comment]
@mschoettle Are all the data you have for QC/MTL found in the following repo: https://github.com/jeremymoreau/covid19mtl/tree/master/app/data/sources

If so, I can just download the whole thing myself and add/replace data as necessary. In fact, it looks like you have some older Quebec data that I would be interested in adding anyway.

**Comment by mschoettle** [comment]
Yes, exactly. Perfect, just keep in mind that the date of the directory is the date for which the data is, not the retrieval date. I.e., just shifted by one day.

And you can find the mapping of the file names here: https://github.com/jeremymoreau/covid19mtl/blob/master/refreshdata.py#L51-L66

**Comment by jeanpaulrsoucy** [comment]
Merged into #182.
